rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/event/RMInitialState.java

```diff
@@ -55,82 +63,136 @@
-                existNodeSource = true;
-                nodeSources.set(i, stateChangedEvent);
-                break;
-            }
-        }
-        if (!existNodeSource) {
-            nodeSources.add(stateChangedEvent);
+    public void nodeRemoved(RMNodeEvent event) {
+        updateCounter(event);
+        nodeEvents.put(event.getNodeUrl(), event);
+    }
+
+    public void nodeSourceAdded(RMNodeSourceEvent event) {
+        updateCounter(event);
+        nodeSourceEvents.put(event.getSourceName(), event);
+    }
+
+    public void nodeSourceRemoved(RMNodeSourceEvent event) {
+        updateCounter(event);
+        nodeSourceEvents.put(event.getSourceName(), event);
+    }
+
+    public void nodeSourceStateChanged(RMNodeSourceEvent event) {
+        updateCounter(event);
+        nodeSourceEvents.put(event.getSourceName(), event);
+    }
+
+    private void updateCounter(RMEvent event) {
+        latestCounter.set(Math.max(latestCounter.get(), event.getCounter()));
+    }
+
+    /**
+     * Clones current state events, but keep only those events which has counter bigger than provided 'filter'
+     * Event counter can take values [0, +).
+     * So if filter is '-1' then all events will returned.
+     * @param filter
+     * @return rmInitialState where all the events bigger than 'filter'
+     */
+    public RMInitialState cloneAndFilter(long filter) {
+        long actualFilter;
+        if (filter <= latestCounter.get()) {
+            actualFilter = filter;
+        } else {
+            LOGGER.info(String.format("Client is aware of %d but server knows only about %d counter. " +
+                                      "Probably because there was network server restart.",
+                                      filter,
+                                      latestCounter.get()));
+            actualFilter = -1; // reset filter to default  value
         }
+        RMInitialState clone = new RMInitialState();
+
+        clone.nodeEvents = newFilteredEvents(this.nodeEvents, actualFilter);
+        clone.nodeSourceEvents = newFilteredEvents(this.nodeSourceEvents, actualFilter);
+
+        clone.latestCounter.set(Math.max(actualFilter,
+                                         Math.max(findLargestCounter(clone.nodeEvents.values()),
+                                                  findLargestCounter(clone.nodeSourceEvents.values()))));
+        return clone;
     }
+
+    private <T extends RMEvent> Map<String, T> newFilteredEvents(Map<String, T> events, long filter) {
+        return events.entrySet()
+                     .stream()
+                     .filter(entry -> entry.getValue().getCounter() > filter)
+                     .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()));
+    }
+
+    private <T extends RMEvent> long findLargestCounter(Collection<T> events) {
+        return events.stream().max(Comparator.comparing(RMEvent::getCounter)).get().getCounter();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Call "Optional#isPresent()" before accessing the value. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3655)